arctic/store/version_store.py

```diff
@@ -560,7 +560,7 @@ def write(self, symbol, data, metadata=None, prune_previous_version=True, **kwar
                                                    sort=[('version', pymongo.DESCENDING)])
 
         handler = self._write_handler(version, symbol, data, **kwargs)
-        mongo_retry(handler.write)(self._arctic_lib, version, symbol, data, previous_version, **kwargs)
+        handler.write(self._arctic_lib, version, symbol, data, previous_version, **kwargs)
 
         # Insert the new version into the version DB
         mongo_retry(self._versions.insert_one)(version)
```

Comment:
This can also go wrong @dimosped 

```
[2018-01-09 09:26:32,608] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,607] {dags_common.py:193} INFO - return self.storage(self._write_lib).mongoose[self._write_lib].write(item, data, metadata=metadata)
[2018-01-09 09:26:32,608] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,607] {dags_common.py:193} INFO - File "<decorator-gen-146>", line 2, in write
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,607] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/ahl/diagscore/events.py", line 326, in event
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - return f(*args, **kwargs)
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/ahl/mongo/decorators.py", line 69, in f_retry
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - return arctic.decorators.mongo_retry(f)(*args, **kwargs)
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/arctic/decorators.py", line 50, in f_retry
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - return f(*args, **kwargs)
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/ahl/mongo/mongoose/store/version_store.py", line 224, in write
[2018-01-09 09:26:32,609] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - prune_previous_version=prune_previous_version, **kwargs)
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/arctic/decorators.py", line 50, in f_retry
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - return f(*args, **kwargs)
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/arctic/store/version_store.py", line 566, in write
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - mongo_retry(self._versions.insert_one)(version)
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/arctic/decorators.py", line 50, in f_retry
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,608] {dags_common.py:193} INFO - return f(*args, **kwargs)
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/pymongo/collection.py", line 682, in insert_one
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - session=session),
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/pymongo/collection.py", line 598, in _insert
[2018-01-09 09:26:32,610] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - bypass_doc_val, session)
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/pymongo/collection.py", line 579, in _insert_one
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - _check_write_command_response(result)
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/pymongo/helpers.py", line 208, in _check_write_command_response
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - _raise_last_write_error(write_errors)
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - File "build/bdist.linux-x86_64/egg/pymongo/helpers.py", line 189, in _raise_last_write_error
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - raise DuplicateKeyError(error.get("errmsg"), 11000, error)
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO -DuplicateKeyError: E11000 duplicate key error collection: mongoose_equities.centaur.versions index: _id_ dup key: { : ObjectId('5a54893cedaca1039023b5b1') }
[2018-01-09 09:26:32,611] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO -
[2018-01-09 09:26:32,612] {base_task_runner.py:95} INFO - Subtask: [2018-01-09 09:26:32,609] {dags_common.py:193} INFO - Exception end: ======================
```

Reply:
Please check this quick suggestion for fix:
https://github.com/manahl/arctic/pull/487/commits/999beb34137f048b4b9dd3ddfbfebd3ee22d1acf

I want to keep the retry there so that we don't create new garbage segments if the error can recover with a local retry.

However, DuplicateKeyError is catastrophic and forces all levels of mongo_retry to exit/give-up, hence special handling is required.

If @jamesblackburn  and @bmoscon  like this solution, I will update the rest of the file and add tests.

Reply:
Simply removing the mongo_retry won't do the trick here, bacuase DuplicateKeyError also forces the method level decorator to exit/quit:
https://github.com/manahl/arctic/blob/master/arctic/decorators.py#L54